### PR TITLE
Add depends_on :linux and :macos

### PR DESCRIPTION
**Pull request by sjackman**
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/master/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [x] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/homebrew/pull/49031).
- [x] Have you successfully run `brew tests` with your changes locally?

-----

Allow a formula to declare whether it depends on macOS or Linux.

### CODE REVIEW

File: Library/Homebrew/dependency_collector.rb [hunk 1]
```diff
@@ -108,7 +108,8 @@ def parse_symbol_spec(spec, tags)
     case spec
     when :x11        then X11Requirement.new(spec.to_s, tags)
     when :xcode      then XcodeRequirement.new(tags)
-    when :macos      then MinimumMacOSRequirement.new(tags)
+    when :linux      then LinuxRequirement.new
+    when :macos      then tags.empty? ? MacOSRequirement.new : MinimumMacOSRequirement.new(tags)
```
Comment (MikeMcQuaid) [inline Library/Homebrew/dependency_collector.rb]: make `MacOSRequirement.new` handle the `tags.empty?` case instead of having the meaning overloaded.

**Comment by ilovezfs** [comment]
How do you specify _both_ that macOS is required _and_ that a minimum version is required? My understanding is `depends_on :macos => :el_capitan` is a no-op on Linux.

**Comment by sjackman** [comment]
To specify that both that macOS is required and that a minimum version is required:
```
depends_on :macos
depends_on MinimumMacOSRequirement => :el_capitan
```

**Comment by ilovezfs** [comment]
That seems like pretty yucky syntax to me.

File: Library/Homebrew/dependency_collector.rb (hunk 1 above)
Comment (sjackman) [reply Library/Homebrew/dependency_collector.rb]: Done.

File: Library/Homebrew/requirements/minimum_macos_requirement.rb [hunk 2]
```diff
@@ -3,18 +3,32 @@
...
-  def initialize(tags)
-    @version = MacOS::Version.from_symbol(tags.first)
+  def initialize(tags = [])
+    @version = MacOS::Version.from_symbol tags.first unless tags.empty?
     super
   end
 
-  satisfy(build_env: false) { MacOS.version >= @version }
+  satisfy(build_env: false) do
+    if @version
```
Comment (MikeMcQuaid) [inline Library/Homebrew/requirements/minimum_macos_requirement.rb]: `next false if !OS.mac? || !@version`

File: Library/Homebrew/requirements/minimum_macos_requirement.rb [hunk 3]
```diff
@@ -3,18 +3,32 @@
 class MinimumMacOSRequirement < Requirement
```
Comment (MikeMcQuaid) [inline Library/Homebrew/requirements/minimum_macos_requirement.rb]: Might be worth renaming this now to be consistent, what do you think?

File: Library/Homebrew/requirements/minimum_macos_requirement.rb (hunk 2 above)
Comment (MikeMcQuaid) [reply Library/Homebrew/requirements/minimum_macos_requirement.rb]: May be worth taking the same approach below, too.

File: Library/Homebrew/requirements/minimum_macos_requirement.rb (hunk 3 above)
Comment (sjackman) [reply Library/Homebrew/requirements/minimum_macos_requirement.rb]: The name of the class is a component of the DSL. The name `MinimumMacOSRequirement` is used by two formulae in core:
```
nordugrid-arc.rb:  depends_on MinimumMacOSRequirement => :yosemite
phantomjs.rb:  depends_on MinimumMacOSRequirement => :lion
```

Comment (MikeMcQuaid) [reply Library/Homebrew/requirements/minimum_macos_requirement.rb]: We can add it to `compat` and update them to use `depends_on :macos`

Comment (sjackman) [reply Library/Homebrew/requirements/minimum_macos_requirement.rb]: Got it. Would you like to suggest a name?

Comment (MikeMcQuaid) [reply Library/Homebrew/requirements/minimum_macos_requirement.rb]: `MacOSRequirement`?

File: Library/Homebrew/requirements/minimum_macos_requirement.rb (hunk 2 above)
Comment (sjackman) [reply Library/Homebrew/requirements/minimum_macos_requirement.rb]: Done. Thanks for the suggestion, Mike. The logic needed tweaking. When `OS.mac?` is `true` and `@version` is `nil`, the result should be `true`.

File: Library/Homebrew/requirements/minimum_macos_requirement.rb (hunk 3 above)
Comment (sjackman) [reply Library/Homebrew/requirements/minimum_macos_requirement.rb]: I've renamed `MinimumMacOSRequirement` to `MacOSRequirement` and added `MinimumMacOSRequirement` to `compat`.

File: Library/Homebrew/compat/requirements/minimum_macos_requirement.rb [hunk 4]
```diff
@@ -0,0 +1,3 @@
+require "requirements/macos_requirement"
```
Comment (MikeMcQuaid) [inline Library/Homebrew/compat/requirements/minimum_macos_requirement.rb]: Stick this file's contents in `homebrew/Library/Homebrew/compat/requirements.rb`

File: Library/Homebrew/requirements/macos_requirement.rb [hunk 5]
```diff
@@ -0,0 +1,26 @@
+require "requirement"
+
+class MacOSRequirement < Requirement
+  fatal true
+
+  def initialize(tags = [])
+    @version = MacOS::Version.from_symbol tags.first unless tags.empty?
```
Comment (MikeMcQuaid) [inline Library/Homebrew/requirements/macos_requirement.rb]: `MacOS::Version.from_symbol(tags.first)` just to aid visual parsing.

File: Library/Homebrew/requirements/macos_requirement.rb [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+    super
+  end
+
+  satisfy(build_env: false) do
+    next false if !OS.mac? && !@version
+    next true if !OS.mac? || !@version
```
Comment (MikeMcQuaid) [inline Library/Homebrew/requirements/macos_requirement.rb]: This is a little hard to parse above. Can you maybe make it `if`/`else`s? This may contradict what I said earlier: sorry!

File: Library/Homebrew/requirements/macos_requirement.rb [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+    MacOS.version >= @version
+  end
+
+  def message
+    return "macOS is required." if !OS.mac? || !@version
```
Comment (MikeMcQuaid) [inline Library/Homebrew/requirements/macos_requirement.rb]: `if !OS.mac? || !@version` could be a utility method, I think, it's repeated 3 times.

File: Library/Homebrew/requirements/macos_requirement.rb (hunk 6 above)
Comment (sjackman) [reply Library/Homebrew/requirements/macos_requirement.rb]: Factoring out the `minimum_version?` utility method also helped make this logic more clear.

File: Library/Homebrew/requirements/macos_requirement.rb [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+
+class MacOSRequirement < Requirement
+  fatal true
+
+  def initialize(tags = [])
+    @version = MacOS::Version.from_symbol(tags.first) unless tags.empty?
+    super
+  end
+
+  def minimum_version?
```
Comment (MikeMcQuaid) [inline Library/Homebrew/requirements/macos_requirement.rb]: `minimum_version_provided` or `passed` or `specified` something would be nicer to make it clear that this doesn't mean the minimum version is met.

**Comment by MikeMcQuaid** [comment]
Tests failing on Linux (shouldn't assume `MacOS` is specified there): https://travis-ci.org/Homebrew/brew/jobs/285144949

**Comment by sjackman** [comment]
Fixed up.

**Comment by MikeMcQuaid** [comment]
@sjackman Still got 🔴.

**Comment by sjackman** [comment]
😬

**Comment by sjackman** [comment]
Good to merge?

File: Library/Homebrew/requirements/macos_requirement.rb [hunk 9]
```diff
@@ -0,0 +1,29 @@
...
+    @version = MacOS::Version.from_symbol(tags.first) unless tags.empty?
+    super
+  end
+
+  def minimum_version_specified?
+    OS.mac? && @version
+  end
+
+  satisfy(build_env: false) do
+    next OS.mac? || @version unless minimum_version_specified?
```
Comment (MikeMcQuaid) [inline Library/Homebrew/requirements/macos_requirement.rb]: Sorry, last nite: can you split this into two `next` checks and return a boolean value?

Comment (sjackman) [reply Library/Homebrew/requirements/macos_requirement.rb]: Done.

**Comment by MikeMcQuaid** [comment]
Thanks @sjackman and sorry for the nit picking!

**Comment by sjackman** [comment]
No worries. Thanks for merging, Mike.
